### ISSUE TITLE
Vim9: substitution with alternation in pattern fails when range is star

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#6938]
**Describe the bug**

In Vim9 script, a substitution with an alternation in its pattern fails when the range is a star (alias for `'<,'>`).

**To Reproduce**

Run this shell command:

    vim -Nu NONE -S <(cat <<'EOF'
        vim9script
        def Func()
            new
            setline(1, ['xxx foo xxx', 'xxx bar xxx', 'xxx foo xx bar'])
            setpos("'<", [0, 1, 0, 0])
            setpos("'>", [0, 3, 0, 0])
            :*s/\%(foo\|bar\)/baz/g
        enddef
        Func()
    EOF
    )

`E486` is raised:

    E486: Pattern not found: \%(foo|bar\)

`foo` and `bar` are *not* replaced with `baz`.

**Expected behavior**

No error is raised.

`foo` and `bar` *are* replaced with `baz`.

**Environment**

 - Vim version: 8.2 Included patches: 1-1665
 - OS: Ubuntu 16.04.7 LTS
 - Terminal: xterm(359)

**Additional context**

The substitution command should work because according to `:h :vim9`:

> A side effect of `:vim9script` is that the 'cpoptions' option is set to the
> Vim default value, like with: >
>         :set cpo&vim

The default value of `'cpo'` does not include the flag `*`.
And according to `:h cpo-star`, if `'cpo'` does not include the `*` flag, then we can use `*` as an alias for `'<,'>`:

> *cpo-star*
> *	Use ":*" in the same way as ":@".  When not included,
> ":*" is an alias for ":'<,'>", select the Visual area.

---

The issue disappears when `*` is replaced with `'<,'>`:
```vim
vim9script
def Func()
    new
    setline(1, ['xxx foo xxx', 'xxx bar xxx', 'xxx foo xx bar'])
    setpos("'<", [0, 1, 0, 0])
    setpos("'>", [0, 3, 0, 0])
    :'<,'>s/\%(foo\|bar\)/baz/g
enddef
Func()
```
---

The issue disappears at the script level:
```vim
vim9script
new
setline(1, ['xxx foo xxx', 'xxx bar xxx', 'xxx foo xx bar'])
setpos("'<", [0, 1, 0, 0])
setpos("'>", [0, 3, 0, 0])
:*s/\%(foo\|bar\)/baz/g
```
---

The issue disappears when the alternation is removed from the pattern:
```vim
vim9script
def Func()
    new
    setline(1, ['xxx foo xxx', 'xxx bar xxx', 'xxx foo xx bar'])
    setpos("'<", [0, 1, 0, 0])
    setpos("'>", [0, 3, 0, 0])
    :*s/foo/baz/g
    :*s/bar/baz/g
enddef
Func()
```
---

Patch adding a test:
```diff
diff --git a/src/testdir/test_vim9_func.vim b/src/testdir/test_vim9_func.vim
index c80271626..df2a219e4 100644
--- a/src/testdir/test_vim9_func.vim
+++ b/src/testdir/test_vim9_func.vim
@@ -1812,5 +1812,15 @@ def Test_callstack_def()
   endtry
 enddef
 
+def Test_star_range()
+  new
+  setline(1, ['xxx foo xxx', 'xxx bar xxx', 'xxx foo xxx bar'])
+  setpos("'<", [0, 1, 0, 0])
+  setpos("'>", [0, 3, 0, 0])
+  :*s/\%(foo\|bar\)/baz/g
+  getline(1, '$')->assert_equal(['xxx baz xxx', 'xxx baz xxx', 'xxx baz xxx baz'])
+  bwipe!
+enddef
+
 
 " vim: ts=8 sw=2 sts=2 expandtab tw=80 fdm=marker
 ```

**Comment by lacygoill** [comment]
> E486 is raised:
>
> E486: Pattern not found: \%(foo|bar\)

For some reason, the backslash in the alternation is wrongly removed:

    \%(foo|bar\)
          ^
          ✘

**Comment by lacygoill** [comment]
Fixed by [8.2.1679](https://github.com/vim/vim/releases/tag/v8.2.1679).
